reconcile/aws_saml_roles/integration.py

```diff
@@ -0,0 +1,160 @@
+from reconcile.utils.aws_api import AWSApi
+from reconcile.utils.defer import defer
+from reconcile.utils.disabled_integrations import integration_is_enabled
+from reconcile.utils.runtime.integration import (
+    PydanticRunParams,
+    QontractReconcileIntegration,
+)
+from reconcile.utils.semver_helper import make_semver
+from reconcile.utils.terraform_client import TerraformClient
+from reconcile.utils.terrascript_aws_client import TerrascriptClient
+
+QONTRACT_INTEGRATION = "aws-saml-roles"
+QONTRACT_INTEGRATION_VERSION = make_semver(1, 0, 0)
+
+
+class AwsSamlRolesIntegrationParams(PydanticRunParams):
+    thread_pool_size: int = 10
+    print_to_file: str | None = None
+    enable_deletion: bool = False
+    # integration specific parameters
+    saml_idp_name: str
+    max_session_duration_hours: int = 1
+    account_name: str | None = None
+
+    @validator("max_session_duration_hours")
+    def max_session_duration_range(cls, v: str | int) -> int:
+        if 1 <= int(v) <= 12:
+            return int(v)
+        raise ValueError("max_session_duration_hours must be between 1 and 12 hours")
+
+
+class AwsSamlRolesIntegration(
+    QontractReconcileIntegration[AwsSamlRolesIntegrationParams]
+):
+    """Manage the SAML IAM roles for all AWS accounts with SSO enabled"""
+
+    @property
+    def name(self) -> str:
+        return QONTRACT_INTEGRATION
+
+    def get_early_exit_desired_state(
+        self, query_func: Callable | None = None
+    ) -> dict[str, Any]:
+        """Return the desired state for early exit."""
+        if not query_func:
+            query_func = gql.get_api().query
+        return {
+            "aws_groups": [c.dict() for c in self.get_aws_groups(query_func)],
+        }
+
+    def get_aws_accounts(
+        self, query_func: Callable, account_name: str | None = None
+    ) -> list[AWSAccountV1]:
+        """Get all AWS accounts."""
+        data = aws_accounts_query(query_func)
+        return [
+            account
+            for account in data.accounts or []
+            if integration_is_enabled(self.name, account)
+            and (not account_name or account.name == account_name)
+        ]
+
+    def get_aws_groups(
+        self, query_func: Callable, account_name: str | None = None
+    ) -> list[AWSGroupV1]:
+        """Get all AWS groups with SSO enabled."""
+        data = aws_groups_query(query_func)
+        return [
+            group
+            for group in data.aws_groups or []
+            if integration_is_enabled(self.name, group.account)
+            and (not account_name or group.account.name == account_name)
+            and group.account.sso
+            and group.roles
+            and group.policies
+            and any(role.users for role in group.roles)
+        ]
+
+    @defer
+    def run(self, dry_run: bool, defer: Callable | None = None) -> None:
+        """Run the integration."""
+        gql_api = gql.get_api()
+        settings = queries.get_app_interface_settings()
```

Comment:
why queries.py?

Reply:
Do you think we can extend `TerrascriptClient` to accept secret_reader? My understanding is settings is only used to instantiate a SecretReader. We do a similar thing for AWSApi to remove the need for settings query https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/utils/aws_api.py#L100

Reply:
Sth in the lines of https://github.com/app-sre/qontract-reconcile/pull/4063

Reply:
I will merge the above first things tomorrow :)

Reply:
Updated. Thanks, Karl.